File: merge_sort.py
'''Functions to implement Merge/sort algorithm'''


def merge_sort(lst):
    """
    Main merge/sort function to call
    takes a list as input and sorts in numerical order
    """
    if len(lst) == 2:
        if lst[0] > lst[1]:
            return [lst[1], lst[0]]
        else:
            return lst
    elif len(lst) <= 1:
        return lst
    else:
        list_a = merge_sort(lst[:len(lst)//2])
        list_b = merge_sort(lst[len(lst)//2:])
        output_list = merge(list_a, list_b)
        return output_list

def merge(list_a, list_b):
    '''merge subroutine called from merge_sort
    takes two sorted sub-lists and merges them into one'''
    output_list = []
    list_a_position = 0
    list_b_position = 0
    while list_a_position + list_b_position < len(list_a) + len(list_b):
        if list_a[list_a_position] < list_b[list_b_position]:
            output_list.append(list_a[list_a_position])
            list_a_position += 1
            if list_a_position == len(list_a):
                output_list.extend(list_b[list_b_position:])
                list_b_position = len(list_b)
        else:
            output_list.append(list_b[list_b_position])
            list_b_position += 1
            if list_b_position == len(list_b):
                output_list.extend(list_a[list_a_position:])
                list_a_position = len(list_a)
    return output_list

File: merge_sort.R-merge_sort---function-lst-len---length-lst-if-len-1-return-lst-else-middle---floor-len/merge_sort.R

merge_sort <- function(lst) {
    len <- length(lst)
    if (len <= 1) {
        return (lst)
    } else {
        middle <- floor(len/2)
        list_a <- merge_sort(lst[1:middle])
        list_b <- merge_sort(lst[(middle + 1):len])
        output_list <- merge(list_a, list_b)
        return (output_list)
    }
}

merge <- function(list_a, list_b) {
    output_list <- c()
    while (length(list_a) > 0 && length(list_b) > 0) {
        if (list_a[1] < list_b[1]) {
            output_list <- c(output_list, list_a[1])
            list_a <- list_a[-1]
        } else {
            output_list <- c(output_list, list_b[1])
            list_b <- list_b[-1]
        }
    }
    if (length(list_a) > 0) {
        output_list <- c(output_list, list_a)
    }
    if (length(list_b) > 0) {
        output_list <- c(output_list, list_b)
    }
    return (output_list)
}
